src/Components/Web/src/Virtualization/Virtualize.cs

```diff
@@ -0,0 +1,264 @@
+
+            builder.OpenElement(4, "div");
+            builder.AddAttribute(5, "style", GetSpacerStyle(itemsAfter));
+            builder.AddElementReferenceCapture(6, elementReference => _spacerAfter = elementReference);
+
+            builder.CloseElement();
+        }
+
+        private string GetSpacerStyle(int itemsInSpacer)
+            => $"height: {itemsInSpacer * ItemSize}px;";
+
+        void IVirtualizeJsCallbacks.OnBeforeSpacerVisible(float spacerSize, float containerSize)
+        {
+            _itemsBefore = CalcualteItemDistribution(spacerSize, containerSize);
+
+            StateHasChanged();
+        }
+
+        void IVirtualizeJsCallbacks.OnBottomSpacerVisible(float spacerSize, float containerSize)
+        {
+            var itemsAfter = CalcualteItemDistribution(spacerSize, containerSize);
+
+            _itemsBefore = Math.Max(0, _itemCount - itemsAfter - _itemsVisible);
+
+            if (_itemsBefore + _visibleItemCapacity > _loadedItems.Count)
+            {
+                _fetchTask = FetchItems();
+            }
+
+            StateHasChanged();
+        }
+
+        private int CalcualteItemDistribution(float spacerSize, float containerSize)
+        {
+            _visibleItemCapacity = (int)Math.Ceiling(containerSize / ItemSize) + 2;
+            _itemsVisible = Math.Clamp(_itemCount, 0, _visibleItemCapacity);
+
+            return Math.Max(0, (int)Math.Floor(spacerSize / ItemSize) - 1);
+        }
+
+        private async Task FetchItems()
+        {
+            // Cancel the previous fetch, if it exists.
+            _fetchCts?.Cancel();
+
+            // Wait for the task to complete. If it fails due to an exception, the exception will bubble-up
+            // through `OnAfterRenderAsync`.
+            if (!_fetchTask.IsCanceled)
+            {
+                await _fetchTask;
+            }
+
+            var placeholderItems = _itemsBefore + _visibleItemCapacity - _loadedItems.Count;
+
+            // There's a chance that no more items need to be fetched after the previous task completes.
+            if (placeholderItems <= 0)
+            {
+                return;
+            }
+
+            _fetchCts = new CancellationTokenSource();
+
+            var result = await _itemsProvider(new ItemsProviderRequest(_loadedItems.Count, placeholderItems, _fetchCts.Token));
+
+            _itemCount = result.TotalItemCount;
+            _loadedItems.AddRange(result.Items);
+        }
+
+        private Task<ItemsProviderResult<TItem>> DefaultItemsProvider(ItemsProviderRequest request)
+        {
+            return Task.FromResult(new ItemsProviderResult<TItem>(
+                Items.Skip(request.StartIndex).Take(request.Count).ToList(),
+                Items.Count));
+        }
+
+        private RenderFragment DefaultPlaceholder(int index) => (builder) =>
+        {
+            builder.OpenElement(0, "div");
+            builder.AddAttribute(1, "style", $"height: {ItemSize}px !important;");
```

Comment:
Is `!important` required? A `style` attribute overrides any CSS-declared styling regardless of `!important` (or so I thought).